NativePromise unnecessarily calls crossThreadCopy with the completion promise

jyavenard opened the pull request:
#### 0e9b295049f8da9354d232b8ab175b30ea65a0cb
<pre>
NativePromise unnecessarily calls crossThreadCopy with the completion promise
<a href="https://bugs.webkit.org/show_bug.cgi?id=262660">https://bugs.webkit.org/show_bug.cgi?id=262660</a>
rdar://116489817

Reviewed by Youenn Fablet.

Avoid copying the data when not needed by separating the code path from
the standard resolve/reject from chained promise resolution.

Add API test to verify that the value isn&apos;t incorrectly moved for non-exclusive
promise.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/269075@main">https://commits.webkit.org/269075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1706b7f5a75c6c8b3fc8b8da463b5bb8c4ed9c2d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21531 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22584 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23397 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19951 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21774 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/25139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22084 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21142 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21759 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21397 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18668 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24248 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18595 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19518 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25822 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/18733 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19675 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19725 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23672 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20911 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17215 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24936 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19525 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5898 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5142 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23764 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26202 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20111 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5727 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1033,11 +1037,11 @@ class NativePromise<ResolveValueT, RejectValueT, options>::Producer final : publ
 
-    explicit operator bool() const { return m_promise && m_promise->isResolvedOrRejected(); }
+    explicit operator bool() const { return m_promise && m_promise->isSettled(); }
     bool isNothing() const
```
s/isNothing/isNotSettled/

jyavenard replied:
I'm going to leave that for now, there's too much use of it and it's purely for assertion purposes

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -420,6 +420,16 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
     }
 
+    template<typename SettleValueType>
+    void settleChainedPromise(SettleValueType&& result, const Logger::LogSiteIdentifier& site)
```
Using twice settleChainedPromise name for two different methods makes things more difficult to read than it should.
How about renaming this one to `settleAsChainedPromise`.

tadeuzagallo committed:
[WGSL] Array length needs to divide size received from the API
https://bugs.webkit.org/show_bug.cgi?id=262793
rdar://116583287

Reviewed by Mike Wyrzykowski.

For runtime-sized arrays, since the API doesn't know the type of the array element
(and consequently its size), it just passes the size of the buffer in bytes. The
generate code for `arrayLength(x)` must then include the division by the size of
the array's element type.

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::getPacking):

Canonical link: https://commits.webkit.org/269072@main

anttijk committed:
[IFC][Ruby][Integration] Propage style to anomymous inlines
https://bugs.webkit.org/show_bug.cgi?id=262877
rdar://116668321

Reviewed by Alan Baradlay.

Ruby uses anonymous inlines. They need style propagation similar to blocks.

* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::styleDidChange):

Propage similarly to RenderBlock.

* Source/WebCore/rendering/updating/RenderTreeBuilderRuby.cpp:
(WebCore::RenderTreeBuilder::Ruby::findOrCreateParentForStyleBasedRubyChild):

Take care to call initializeStyle() for the anonymous renderers.

Canonical link: https://commits.webkit.org/269073@main

alanbaradlay committed:
[IFC][Cleanup] TextUtil:computedExpansions should resizeToFit to range size
https://bugs.webkit.org/show_bug.cgi?id=262871

Reviewed by Antti Koivisto.

...instead of the run list size.

* Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
(WebCore::Layout::Line::applyExpansionOnRange):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::computedExpansions):

Canonical link: https://commits.webkit.org/269074@main

jyavenard committed:
NativePromise unnecessarily calls crossThreadCopy with the completion promise
https://bugs.webkit.org/show_bug.cgi?id=262660
rdar://116489817

Reviewed by Youenn Fablet.

Avoid copying the data when not needed by separating the code path from
the standard resolve/reject from chained promise resolution.

Add API test to verify that the value isn't incorrectly moved for non-exclusive
promise.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/269075@main
